### ISSUE TITLE
Add interface to CLAW

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#114]
We need to be able to apply CLAW transformations to user-supplied kernels. In this Issue we will start to add the necessary wrappers and documentation.

**Comment by arporter** [comment]
In contrast to PSyclone's native transformations which operate on the
PSy AST, CLAW operates on Fortran source code. In this case this will
be one or more of the kernels called from the PSy layer.
PSyclone already has the location of these kernels since the meta-data
they contain must be parsed. This location information must now also be passed
to the transformation script in case it wishes to modify the kernels.

Naming of the transformed kernel - just name mangle and output it in the
same location as the generated psy code? However, LFRic has naming
rules which means the name of the kernel is intimately linked with the
name of the file which contains it. This means that if we name mangle
then we have to do this consistently through the kernel file (i.e.
the module, type declaration and kernel subroutine will all have to be
renamed). If we don't change the name of the kernel then making sure
we get the right kernel at compile time becomes the problem of the build system so
long as we don't over-write the original kernels. The latter problem
can be avoided by allowing the output directory to be specified as an
argument to PSyclone. This will then have to be passed down to the 
`trans()` function as an additional argument. Therefore, the proposed interface to `trans()` is now:

    def trans(psy, kernel_search_path, out_dir):
        '''
        :param psy: root of PSy AST
        :param kernel_search_path: List of directories to search for kernels
        :param out_dir: directory in which to put any transformed kernels
        '''

**Comment by rupertford** [comment]
There is another practical issue here. If PSyclone is run separately for two different algorithm files which each make use of the same kernel then what do we do? I guess we could probably just say it is a build problem. However to help the build system we might want to have various configuration options for the case where an existing output file is found I.e. 1: overwrite, 2: keep original 3: raise an exception.

There might be a use case where the hpc expert wants to create a modified version of the kernel but also keep the original (or some other variant). In this case I'm not sure it can be done though different directories due to the module names clashing, so we might need to be able to create a unique file and module name that is also used by PSyclone.

**Comment by rupertford** [comment]
OK, I'm going to argue against the proposed 2nd argument to trans() and suggest it should be a dictionary of Kernel AST's, indexed by module name. We then need to decide if each should include its original file name, or whether we create the file name from the module name.

Thinking about Andy's previous comment, it is true that CLAW reads from a fortran file and writes to a fortran file. However, it actually uses OMNII to read the file and output that as an AST (OMNII Fortran IR) which is then modified by the CLAW transformations. So I think of CLAW as being something that modifies a fortran IR and happens to input from a fortran file and output to a fortran file.

PSyclone already has the kernel code in the fparser AST so if these AST's were compatible we could simply pass the fparser AST's in. To spell it out, if we eventually get fparser to be able to output the OMNII Fortran IR then should pass that into the transformation, as this avoids us having to parse kernels twice.

OK, so rather than propose we change the API at a later date I suggest that we should go with the AST solution now (rather than the input directory solution). We could make it look like the AST is read directly by our Python CLAW interface, even though in practice it would need to be turned back into fortran and then parsed by OMNII.

**Comment by arporter** [comment]
I like the suggestion of having the API take ASTs instead of a search path. I was hoping to avoid the additional complexity of having to re-name a kernel but it is probably the right solution.

**Comment by arporter** [comment]
`KernelType._ast` points to the fparser AST for the module containing the kernel code.

**Comment by arporter** [comment]
Given the need to re-name a transformed kernel, the PSy-layer will also need to be updated with the new kernel name. Therefore, the CLAW interface will need to be passed a pointer to the relevant invoke in the PSy layer. However, if a PSy object contains several invokes which make use of the same kernels and we want all of those kernels transformed in the same way then we will end up with duplicated work and duplicated files. We could pass in a list of invokes as well as a list of kernels. Each of the kernels would only be transformed once but would then be used in any/all of the supplied invokes that called it.

**Comment by arporter** [comment]
Unlike fparser, OMNI expects to be able to resolve all variable types when parsing a source file. i.e. it expects to be able to find a `.xmod` file for each module that is `use`d by a kernel file.

**Comment by arporter** [comment]
It is only necessary to run the OMNI front-end on a module to generate the `.xmod` file. However, this will have to be done for all 'infrastructure' modules used by kernels and the location of the resulting `.xmod` files passed in to the CLAW interface?

**Comment by arporter** [comment]
Even though our kernels are typically very Fortran77-like, at the very least they will probably use kind-type parameters and they will have been defined in some other module. I therefore don't think there's any alternative to passing in an "xmod search path" to the CLAW interface.

**Comment by arporter** [comment]
Valentin has confirmed that the OMNI front-end requires the .xmod files for any dependencies.
Next stage is to get a working example with the proposed interface and check that we've not overlooked anything.

**Comment by rupertford** [comment]
As CLAW is being used to optimise/modify kernel code then we will only need to worry about xmod files for LFRic infrastructure calls. We can therefore create such xmod files once and store them in a known location. If we know the location then we can keep the location in a config file and use that config file in the script rather than passing in a location via the API. Would that work?

On a separate note, if we are compiling the infrastructure into xmod files this is similar to the compilation tests. Perhaps we could use the same solution for both somehow e.g. both use the cut-down LFRic, or both use the real LFRic?

**Comment by rupertford** [comment]
Another comment I'm afraid ...

Should the CLAW interface be responsible for updating the PSy code? An alternative would be for the PSy modification part of the script to extract (or be passed) the new name once CLAW has updated a kernel file. This would allow decisions to be made for kernel name changes at the PSy level which is probably the right place given the potential for the same kernel to be called in multiple invokes (or even multiple places within an invoke).

**Comment by arporter** [comment]
Running the compilation tests with the OMNI compiler reveals some problems. I may have to generate some OMNI issues...

**Comment by arporter** [comment]
Currently my only example is for a GOcean kernel. However, I can't build the GOcean infrastructure with Omni because it (rightly) complains that it has name clashes with Fortran intrinsics. I've created an issue to deal with this but am waiting to see whether Joerg has already fixed this.
Next step is therefore to add an LFRic example.

**Comment by hiker** [comment]
I vaguely remember having problems with max or min (which I assume is what you have a problem with as well ... I mentioned it in https://github.com/stfc/fparser/pull/49#discussion_r145666455). I don't remember the details (I only committed the fixed version), but iirc it was related to an ```use argument_mod``` without ```only``` (which then made min available in the subroutine). I think the fix was to use the dl_esm repo, which fixes some of the psyclone issues I had.

So in short, sorry, I never had a fix for that.

**Comment by arporter** [comment]
OK @hiker, that's fine - I just didn't want to duplicate your work :-)

**Comment by arporter** [comment]
Have added an LFRic example. Both the LFRic and GOcean examples sort of run. The former gets an exception when CLAW runs the OMNI backend and the latter gets a seg-fault in the OMNI frontend. Location of xmod files and CLAW python module are now picked-up from the config file.

**Comment by arporter** [comment]
The LFRic problem was down to the fact that I wasn't providing the name of an output xml file (to take the transformed IR) to Claw. I think Claw should generate one if the user hasn't supplied one (since it is only to pass the transformed IR to the omni backend) but that's not something for this issue. We now generate a 'transformed' LFRic kernel (albeit we haven't actually manipulated it yet).

**Comment by arporter** [comment]
Next stage is to properly handle kernel renaming and update the LFRic example so that it actually does something (maybe insert an OpenACC directive?).

**Comment by arporter** [comment]
Some basic unit tests are now implemented. These do not require Claw in order to run. However, I think we will require some functional tests that do involve running Claw for real. I envisage adding a "--with_claw" or similar flag to the py.test command line, similar to "--compile".

**Comment by arporter** [comment]
I've narrowed-down the issue with the GOcean Claw example to a problem with the Omni frontend when a type contains type which contains a procedure. I've posted about this in the Omni google group: https://groups.google.com/forum/#!forum/omni-compiler

It may be that a stripped-down version of dl_esm_inf will allow us to workaround this problem but I think that can wait for the Issue in which we add compilation testing for the GOcean API.

**Comment by arporter** [comment]
We can't use `claw.wani.transformation.ll.loop.LoopInterchange` directly because it expects the target loops to be identified using CLAW directives. However, it is implemented using `claw.tatsu.primitive.Loop` and `claw.tatsu.xcodeml.abstraction.NestedDoStatement`. I was able to `import` these into the PSyclone script and use them to perform a loop interchange.

**Comment by arporter** [comment]
After an initial review of the PR associated with this issue, we have decided NOT to interface between PSyclone and CLAW directly. Instead PSyclone will modify the XCodeML IR of a kernel by inserting CLAW directives. CLAW will then be run on the resulting XCodeML (outside of PSyclone probably). Since this is a very different approach I'm closing this issue and will create a new one.
